File: Interviews/GFG/Daily-Problems/2022/October/Day-19/Hamiltonian-Path.cpp
class Solution {
public:
    bool check(int N, int M, vector<vector<int>> Edges) {
        bool adj[N][N], cache[N][1 << N];
        memset(adj, 0, sizeof(adj));
        memset(cache, 0, sizeof(cache));

        for (auto e : Edges) {
            adj[e[0] - 1][e[1] - 1] = adj[e[1] - 1][e[0] - 1] = 1;
        }

        for (int i = 0; i < N; i++) {
            cache[i][1 << i] = 1;
        }

        for (int i = 0; i < (1 << N); i++) {
            for (int j = 0; j < N; j++) {
                if ((i >> j) & 1) {
                    for (int k = 0; k < N; k++) {
                        if (j != k and ((i >> k) & 1) and adj[j][k] and cache[k][i ^ (1 << j)]) {
                            cache[j][i] = 1;
                            break;
                        }
                    }
                }
            }
        }

        for (int i = 0; i < N; i++) {
            if (cache[i][(1 << N) - 1] == 1) {
                return true;
            }
        }

        return false;
    }
};

File: Interviews/GFG/Daily-Problems/2022/October/Day-19/Hamiltonian-Path.py
class Solution:
    def check(self, N, M, Edges): 
        adj, cache = [[False for i in range(N)] for j in range(N)], [[False for i in range(1 << N)] for j in range(N)]
        for e in Edges:
            adj[e[0] - 1][e[1] - 1] = adj[e[1] - 1][e[0] - 1] = True
        for i in range(N):
            cache[i][1 << i] = True
        
        for i in range(1 << N):
            for j in range(N):
                if ((i >> j) & 1):
                    for k in range(N):
                        if j != k and (i >> k) & 1 and adj[j][k] and cache[k][i ^ (1 << j)]:
                            cache[j][i] = True
                            break
        
        for i in range(N):
            if cache[i][(1 << N) - 1]:
                return True
        
        return False